Improve WritablePolygon2D and WritablePolyline

ctrueden opened the pull request:
This branch makes the following improvements:

* `Polygon2D` and `Polyline` common API is now inherited from `Polyshape` interface.
* `WritablePolygon2D` and `WritablePolyline` common API is now inherited from `WritablePolyshape` interface. The `WritablePolygon2D#addVertex(int, double[])` method is deprecated in favor of `WritablePolyshape#addVertex(int, RealLocalizable)`.
* `DefaultWritablePolygon2D#addVertex`, `DefaultWritablePolyline#addVertex` and `DefaultWritableRealPointCollection#addPoint` now update the bounds based on the new point only, avoiding an O(n) bounds rebuild. (`RealPointSampleListWritableRealPointCollection` did not suffer from this issue.)
* `WritablePolyshape` gained an `addVertices` method for adding many vertices at once. This bulk operation enables the default implementations to perform better, expanding their internal storage once as appropriate, rather than doing repeated incremental expansions.
* Add methods to `GeomMaths` to compare `Localizable` and `RealLocalizable` positions. (Maybe these should go into `net.imglib2.util.Util`? What do others think?)
* The `WritablePolygon2DTest` code has been improved a bit.

Note that `WritableRealPointCollection` still does not have `addPoints` or `removePoints` methods; I considered doing them as part of this PR, but I have no immediate use case, so I've held off for now.

With these changes, it will be more feasible to pass `WritablePolyshape` objects as outputs into which calculated points can be injected. In particular, it will become feasible to recast ImageJ Ops's geometry operations such as `geom.convexHull(Polygon2D)` as computer ops rather than function ops.

awalter17 commented:
@ctrueden Thank you for your work on this (and cleaning up some of my tests)!

> Add methods to `GeomMaths` to compare `Localizable` and `RealLocalizable` positions. (Maybe these should go into `net.imglib2.util.Util`? What do others think?)

I think these methods should go in `net.imglib2.util.Util`, because they seem more general than `GeomMaths`.

Aside from that though, I think these changes are wonderful and (pending a decision on the above) are good to merge. Thoughts @tpietzsch?

ctrueden commented:
Rebased over the latest master. I also added a commit cleaning up the `GeomMasks` javadoc. And I removed the `GeomMaths.locationsEqual` methods in favor of `Util.locationsEqual` from imglib/imglib2#211. This PR depends on imglib/imglib2#211 being merged and released first.

imagejan commented:
The build on `master` failed because of a typo in the javadoc [here](https://github.com/imglib/imglib2-roi/blob/12b0ddcfa55206a46b9c5b6a889a9581dc72b7ab/src/main/java/net/imglib2/roi/geom/GeomMasks.java#L262): `@Link` -> `@link`

ctrueden commented:
Fixed with 07f5716162c1c95f81389a9d8df9819d17f227d7
